### ISSUE TITLE
Use offline dictionary files

### ISSUE THREAD

**Issue by GrimPixel** [jzohrab/lute-v3#228]
**Is your feature request related to a problem? Please describe.**

Offline dictionary files are not supported.

**Describe the solution you'd like**

Support offline dictionary files, at least TSV.

**Additional context**

There is a repository for creating TSV dictionaries: https://codeberg.org/GrimPixel/Dictionary_Tools

**Comment by jzohrab** [comment]
Related to #5 and #6 .

This would also have to include:

- how to (generate?) and install the dictionaries
- how to map terms to the dictionaries
- documentation

**Comment by GrimPixel** [comment]
My thoughts: place dictionary files in a directory, the program will then read them. Some plug-ins can generate non-lemma forms from the lemma; the term can be searched in both the lemma and non-lemma parts.

**Comment by jzohrab** [comment]
Yes, of course :-)  Cheers!  I was only listing some other things that would need to be thought about for this.

This is a good idea, it's exactly what I need sometimes.  I have a Mac and so would like to use the built-in Mac dictionary.  Thanks GP 👋

**Comment by jzohrab** [comment]
Copied from #452:

**Is your feature request related to a problem? Please describe.**
I'm coming from trying out LWT which seems like it's not under active development and a bit dated at this point. I didn't like that to add terms, it basically created a popup to a dictionary and you'd have to manually enter the translation.

**Describe the solution you'd like**
Freedict (https://freedict.org/downloads/) has a number of free offline dictionaries that could be used for prepopulation if there is an issue finding a free dictionary API (maybe there is one, I don't know). 

**Describe alternatives you've considered**
Typing everything in manually becomes cumbersome. I could cobble together my vocab plus some A1+A2 basic word lists for import via CSV but it would take a long time and is overwhelming.

**Comment by jzohrab** [comment]
Copied from #521 :

**Is your feature request related to a problem? Please describe.**

Hello! I'm exploring Lute coming from [Vocab Tracker](https://www.vocabtracker.com), and one thing I miss is instant translations/definitions. 

As I understand it, in Lute, there are no pre-filled definitions, and you have to copy-paste your own definitions from an external dictionary. In my use case, I simply determine how well I know a word's meaning, and assign it a level. I do not intend to bother with assigning data to each word within Lute itself.

However, when you click on a word, loading an online dictionary for each new word can take a second or two (on my connection, at least), and that's enough to break my immersion when reading.

**Describe the solution you'd like**

Allow installing an offline .tsv dictionary (such as the ones used by [LinguaCafe](https://github.com/simjanos-dev/LinguaCafe/releases/tag/dictionaries)), to provide instant definitions for selected words, while in the meantime also loading the embedded online dictionary (if so configured).

**Describe alternatives you've considered**

I'm not sure what solution Vocab Tracker uses to be so efficient, but that! I think it uses some translator service, but whatever it is, it loads meanings instantly when selecting words or even phrases.

**Additional context**

That's all! Thank you!

**Comment by jzohrab** [comment]
Hi @GrimPixel and @sabretou - a few questions about the TSV file idea:

- is the format for such files consistent, or can they be made consistent?
- is a dictionary completely contained in a single tsv file, or is it spread among a few files?

Cheers and thanks :wave:

**Comment by sabretou** [comment]
>     * is the format for such files consistent, or can they be made consistent?

I'm not sure about this one as an end-user. 😅

Lingua Cafe for instance uses tabfile .tsv files [that appear to be from Vuizur's repository](https://github.com/Vuizur/Wiktionary-Dictionaries). That should be very comprehensive for each language.

Vuizur themself appears to have received their data [from Kaikki.org](https://kaikki.org/index.html), which appears to use JSON, but notably also has data from non-English Wiktionaries, therefore allowing more language combinations.

Perhaps that helps?

>     * is a dictionary completely contained in a single tsv file, or is it spread among a few files?

In the case of Vuizur's files, yes, a dictionary is completely contained in a single tsv file.

**Comment by jzohrab** [comment]
Added Help-Wanted tag as this feels like it would be extremely useful, and could be investigated by anyone interested.  If anyone could help provide a simple stub implementation, or even a simple demo of accessing/consuming a local dict file of any format, that would be great.  o/

**Comment by yethu** [comment]
Is this still a desired feature? I recently started using Lute to learn German and I think offline dictionaries would be a great feature. I am interested in working on this.

**Comment by jzohrab** [comment]
(Apologies for the delay in responding, I've had some vision issues, [detail here](https://jzohrab.github.io/lute/2026/04/04/lute-dev-stopped/))

Hi @yethu - thanks very much, I believe that some people would find it useful :-)  There are quite a few gotchas though, I think!  e.g., offhand:

- where to store the dict files
- how to do lookups, given that Lute may store words differently than they're expected in the dictionary files.  Lute uses a "zero-width-space" character between what it considers word boundaries, which are determined by the parsers.  e.g. in Japanese a word composed of two characters, like "AB", might actually be stored as "A/B", with "/" being the zero-width space.
- The way to interface to these dictionaries would need to be worked out!  e.g., would they be somehow accessed using a special URL string for a dictionary, and intercepted by Lute and sent to its own internal dictionary handler?

Sorry these are rough notes, this is what I can recall from my old musings on this issue.  Cheers!

**Comment by yethu** [comment]
Hi @jzohrab 
Really sorry to hear about your vision problems. Please don't worry about the delay at all, health always comes first. I've actually been tied up with my school work in past few months and I'm finally free to start looking at this now. Thanks for pointing out those gotchas. I will also take issues pointed out by other contributors in the thread in consideration. I'll be working on it slowly but surely, and I'll drop an update here once I make some progress. Take care!
